aten/src/ATen/native/TensorIterator.h

```diff
@@ -20,49 +22,61 @@
 //      .build()
 //
 // [MyKernel.cpp / MyKernel.cu]
-//   cpu_binary_kernel(iter, [](float a, float b) {
+//   binary_kernel(iter, [](float a, float b) {
 //     return a + b;
 //   });
 //
 //   gpu_binary_kernel(iter, []GPU_LAMBDA(float a, float b) -> float {
 //     return a + b;
 //   });
 //
-// This inspired by NumPy's Array Iterator API (NpyIter).
+// Note [Result type computation]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// TensorIterator handles limited mixed-type operations. The result type is
+// computed using promoteTypes on the types of the operands with the following
+// precedence:
+//
+// 1) Tensors with dim 1 or higher
+// 2) Tensors with dim 0 that aren't wrapped numbers (e.g. `tensor(5)`)
+// 3) Tensors with dim 0 that are wrapped numbers (e.g. `5`)
+//
+// So if there are any tensors of dim 1 or higher, then 0-dim tensors do not
+// affect the result type. This behavior was chosen to preserve backwards
+// compatibility and is *likely to change* in the near future.
```

Comment:
Why is this likely to change?

Reply:
Currently, `int_tensor + 2.5` returns an int tensor. This behavior is preserved in this PR. In the future, if we adopt https://github.com/pytorch/pytorch/issues/9515, it will return a floating-point tensor.